src/miner.cpp

```diff
@@ -649,6 +709,93 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
     std::sort(sortedEntries.begin(), sortedEntries.end(), CompareTxIterByEntryTime());
 }
 
+bool BlockAssembler::EvmTxPreapply(const EvmTxPreApplyContext& ctx) {
+    auto txMessage = customTypeToMessage(ctx.txType);
+    auto& txIter = ctx.txIter;
+    auto& evmQueueId = ctx.evmQueueId;
+    auto& metadata = ctx.txMetadata;
+    auto& height = ctx.height;
+    auto& pkgCtx = ctx.pkgCtx;
+    auto& checkedDfTxHashSet = ctx.checkedTXEntries;
+    auto& failedTxSet = ctx.failedTxEntries;
+    auto& failedNonces = ctx.pkgCtx.failedNonces;
+    auto& replaceByFee = ctx.pkgCtx.replaceByFee;
+
+    auto txIters = [](std::map<uint64_t, CTxMemPool::txiter> &iterMap) -> std::vector<CTxMemPool::txiter> {
+        std::vector<CTxMemPool::txiter> txIters;
+        for (const auto& [nonce, it] : iterMap) {
+            txIters.push_back(it);
+        }
+        return txIters;
+    };
+
+    if (!CustomMetadataParse(height, Params().GetConsensus(), metadata, txMessage)) {
+        return false;
+    }
+
+    const auto obj = std::get<CEvmTxMessage>(txMessage);
+
+    CrossBoundaryResult result;
+    const auto txResult = evm_try_prevalidate_raw_tx(result, HexStr(obj.evmTx));
+    if (!result.ok) {
+        return false;
+    }
+
+    auto& evmFeeMap = pkgCtx.feeMap;
+    auto& evmAddressTxsMap = pkgCtx.addressTxsMap;
+
+    const auto addrKey = EvmAddressWithNonce { txResult.sender, txResult.nonce }; 
+    if (auto feeEntry = evmFeeMap.find(addrKey); feeEntry != evmFeeMap.end()) {
+        // Key already exists. We check to see if we need to prioritize higher fee tx
+        const auto& lastFee = feeEntry->second;
+        if (txResult.prepay_fee > lastFee) {
+            // Higher paying fee. Remove all TXs from sender and add to collection to add them again in order.
+            auto addrTxs = evmAddressTxsMap[addrKey.address];
+            for (auto iter: txIters(addrTxs)) {
+                if (RemoveFromBlock(iter)) {
+                    inBlock.erase(iter);
+                }
+            }
+            // Remove all fee entries relating to the address
+            for (auto it = evmFeeMap.begin(); it != evmFeeMap.end();) {
+                const auto& [sender, nonce] = it->first;
+                if (sender == addrKey.address) {
+                    it = evmFeeMap.erase(it);
+                } else {
+                    ++it;
+                }
+            }
+            // Buggy code to fix below:
+            checkedDfTxHashSet.erase(addrTxs[addrKey.nonce]->GetTx().GetHash());
+            addrTxs[addrKey.nonce] = txIter;
+            auto count{addrKey.nonce};
+            for (const auto& [nonce, entry] : addrTxs) {
+                checkedDfTxHashSet.erase(entry->GetTx().GetHash());
+                replaceByFee.emplace(count, entry);
```

Comment:
Similar to PR #2220 - is there a reason to be using count instead of directly using the nonce matching the evm tx when storing descendant evm txs inside this index?

Reply:
We did not have the nonce before and they should have been in nonce order. Will update this.